### ISSUE TITLE
Performance in Elex is more than halved when using winelib dxvk

### ISSUE THREAD

**Issue by shmerl** [doitsujin/dxvk#805]
I noticed a strange thing. Performance in Elex is more than halved when using winelib dxvk, in comparison with dll one. Is that expected?

I don't observe such degradation for example in The Witcher 3. Performance is practically the same there between winelib and dll builds.

dll: 

![elex_dxvk_dll](https://user-images.githubusercontent.com/310927/49693247-70dd6580-fb3b-11e8-960a-f94c2f03377d.jpg)

so (winelib): 

![elex_dxvk_so](https://user-images.githubusercontent.com/310927/49693248-76d34680-fb3b-11e8-8129-69435ccbfa9a.jpg)

Dxvk: latest master (post 0.93).
GPU: AMD Sapphire Pulse Vega 56.
radv: Mesa master, llvm 8.0.

**Comment by doitsujin** [comment]
Any particular reason you use winelib builds with Elex? Are you maybe getting some log spam from wine? Did you build it with compiler optimizations enabled?

**Comment by shmerl** [comment]
No specific reason, just wanted to compare performance. I ran it with:

```
export WINEDEBUG=-all
export DXVK_LOG_LEVEL=none
```
So it shouldn't have log spammed it. Let me check how it was compiled.

**Comment by shmerl** [comment]
I built it as using this (setting some variables before):
```
  meson --cross-file build-wine64.txt --prefix ${dest_dir} "$dxvk_build"
  cd "$dxvk_build"
  meson configure -Dbuildtype=release
  ninja
```
Does release enable compiler optimizations?

**Comment by pchome** [comment]
@shmerl 
> Does release enable compiler optimizations?

`buildtype=release` enable `-O3` optimization level. The flags enabled at this level can vary between GCC versions, see e.g. `enabled by default at -O3` ("General Improvements" section):
* https://gcc.gnu.org/gcc-7/changes.html
* https://gcc.gnu.org/gcc-8/changes.html

You can:
* check compiler versions used for build (Meson usually prints both, during MinGW build)
* modify your `build-wine64.txt` to use `-O2`, and try e.g. `-Dbuildtype=plain`
  or (for newer Meson versions) `-Dbuildtype=custom -Doptimization=2`

Such steps will help to find out is regression caused by `-O3`, by the specific flag, or by different GCC version.

**Comment by kakra** [comment]
The other question is: Did you build the DLL version yourself or did you use the pre-compiled one? Because you didn't mention how you built the other version...

Otherwise, comparing both versions is meaningless: The compiler is probably different, and versions may differ.

**Comment by shmerl** [comment]
I built both myself. I'll run some tests using suggestions to lower optimization level for winelib case.

**Comment by shmerl** [comment]
Compilers used:

* Winelib build:
```
Native C compiler: cc (gcc 8.2.0 "cc (Debian 8.2.0-9) 8.2.0")
Cross C compiler: winegcc (gcc 8.2.0)
Native C++ compiler: c++ (gcc 8.2.0 "c++ (Debian 8.2.0-9) 8.2.0")
Cross C++ compiler: wineg++ (gcc 8.2.0)
```

dll build:

```
Native C compiler: cc (gcc 8.2.0 "cc (Debian 8.2.0-9) 8.2.0")
Cross C compiler: x86_64-w64-mingw32-gcc (gcc 7.3.0)
Native C++ compiler: c++ (gcc 8.2.0 "c++ (Debian 8.2.0-9) 8.2.0")
Cross C++ compiler: x86_64-w64-mingw32-g++ (gcc 7.3.0)
```

So there is that difference already. Let me try newer mingw.

**Comment by shmerl** [comment]
Nope, the kernel doesn't make a difference. Not sure what changed really.

**Comment by kakra** [comment]
@shmerl Maybe you are experiencing the same pitfall I did and during tests something was eating your VRAM, e.g. an open browser with multiple tabs which wasn't open during on of your test cases (or memory usage was different)? The effect was about half the performance here...

**Comment by shmerl** [comment]
Good point, I didn't monitor my VRAM usage. I'm actually not sure how to do that for amdgpu. Are there some good methods?

**Comment by shmerl** [comment]
Ah, this produces something, but probably means pages:
```
sudo cat /sys/kernel/debug/dri/0/amdgpu_vram_mm | pcregrep 'total|usage'
total: 2093056, used 152311 free 1940745
man size:2093056 pages, ram usage:594MB, vis usage:42MB
```

**Comment by shmerl** [comment]
[Created a crude GPU load / VRAM monitor if it's useful for anyone](https://gitlab.freedesktop.org/snippets/540).

**Comment by pchome** [comment]
> Created a crude AMD GPU load / VRAM monitor

https://github.com/clbr/radeontop ?

**Comment by shmerl** [comment]
It tried radeontop, it produces no data for my Vega.

**Comment by doitsujin** [comment]
I can't even get the game to start, so I won't be able to look into it for now.

**Comment by shmerl** [comment]
Is it crashing?

**Comment by doitsujin** [comment]
It shows an unreadable crash dialogue and then proceeds to show an infinite loading screen. Not sure why but I've had this with all wine versions past 3.14, and now even my 3.14 build stopped working.

**Comment by shmerl** [comment]
Hm, just for the record, I'm using 4.0~rc1, but never experienced such crash wither earlier versions as well.

**Comment by kakra** [comment]
@shmerl Did you set CFLAGS or something similar? Or update meson lately? While debugging a build issue here, I noticed that the winelib build no longer respected my CFLAGS set through a makefile (or maybe it never did and I didn't notice). Upon inspection, I found that there was a fix to meson to ignore CFLAGS for cross-builds. Since winelib is implemented as a meson cross-build, that could explain something...
